In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.utils import resample
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn import preprocessing

from datetime import datetime, date

%matplotlib inline

In [2]:
raw_df = pd.read_csv("data.csv")

### Exploratory Data Analysis

In [3]:
# View data
raw_df.head()

UniqueID  disbursed_amount  asset_cost    ltv  branch_id  supplier_id  \
0    420825             50578       58400  89.55         67        22807   
1    537409             47145       65550  73.23         67        22807   
2    417566             53278       61360  89.63         67        22807   
3    624493             57513       66113  88.48         67        22807   
4    539055             52378       60300  88.39         67        22807   

   manufacturer_id  Current_pincode_ID Date.of.Birth Employment.Type  ...  \
0               45                1441      01-01-84        Salaried  ...   
1               45                1502      31-07-85   Self employed  ...   
2               45                1497      24-08-85   Self employed  ...   
3               45                1501      30-12-93   Self employed  ...   
4               45                1495      09-12-77   Self employed  ...   

  SEC.SANCTIONED.AMOUNT  SEC.DISBURSED.AMOUNT  PRIMARY.INSTAL.AMT  \
0                     0                     0                   0   
1                     0                     0                1991   
2                     0                     0                   0   
3                     0                     0                  31   
4                     0                     0                   0   

   SEC.INSTAL.AMT  NEW.ACCTS.IN.LAST.SIX.MONTHS  \
0               0                             0   
1               0                             0   
2               0                             0   
3               0                             0   
4               0                             0   

   DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS  AVERAGE.ACCT.AGE  \
0                                    0         0yrs 0mon   
1                                    1        1yrs 11mon   
2                                    0         0yrs 0mon   
3                                    0         0yrs 8mon   
4                                    0         0yrs 0mon   

   CREDIT.HISTORY.LENGTH  NO.OF_INQUIRIES  loan_default  
0              0yrs 0mon                0             0  
1             1yrs 11mon                0             1  
2              0yrs 0mon                0             0  
3              1yrs 3mon                1             1  
4              0yrs 0mon                1             1  

[5 rows x 41 columns]

In [4]:
raw_df.describe().round(3)

UniqueID  disbursed_amount   asset_cost         ltv   branch_id  \
count  233154.000        233154.000   233154.000  233154.000  233154.000   
mean   535917.573         54356.994    75865.068      74.747      72.936   
std     68315.694         12971.314    18944.781      11.457      69.835   
min    417428.000         13320.000    37000.000      10.030       1.000   
25%    476786.250         47145.000    65717.000      68.880      14.000   
50%    535978.500         53803.000    70946.000      76.800      61.000   
75%    595039.750         60413.000    79201.750      83.670     130.000   
max    671084.000        990572.000  1628992.000      95.000     261.000   

       supplier_id  manufacturer_id  Current_pincode_ID    State_ID  \
count   233154.000       233154.000          233154.000  233154.000   
mean     19638.635           69.028            3396.880       7.262   
std       3491.950           22.141            2238.148       4.482   
min      10524.000           45.000               1.000       1.000   
25%      16535.000           48.000            1511.000       4.000   
50%      20333.000           86.000            2970.000       6.000   
75%      23000.000           86.000            5677.000      10.000   
max      24803.000          156.000            7345.000      22.000   

       Employee_code_ID  ...  SEC.OVERDUE.ACCTS  SEC.CURRENT.BALANCE  \
count        233154.000  ...         233154.000         2.331540e+05   
mean           1549.477  ...              0.007         5.427793e+03   
std             975.261  ...              0.111         1.702370e+05   
min               1.000  ...              0.000        -5.746470e+05   
25%             713.000  ...              0.000         0.000000e+00   
50%            1451.000  ...              0.000         0.000000e+00   
75%            2362.000  ...              0.000         0.000000e+00   
max            3795.000  ...              8.000         3.603285e+07   

       SEC.SANCTIONED.AMOUNT  SEC.DISBURSED.AMOUNT  PRIMARY.INSTAL.AMT  \
count           2.331540e+05          2.331540e+05        2.331540e+05   
mean            7.295923e+03          7.179998e+03        1.310548e+04   
std             1.831560e+05          1.825925e+05        1.513679e+05   
min             0.000000e+00          0.000000e+00        0.000000e+00   
25%             0.000000e+00          0.000000e+00        0.000000e+00   
50%             0.000000e+00          0.000000e+00        0.000000e+00   
75%             0.000000e+00          0.000000e+00        1.999000e+03   
max             3.000000e+07          3.000000e+07        2.564281e+07   

       SEC.INSTAL.AMT  NEW.ACCTS.IN.LAST.SIX.MONTHS  \
count      233154.000                    233154.000   
mean          323.268                         0.382   
std         15553.691                         0.955   
min             0.000                         0.000   
25%             0.000                         0.000   
50%             0.000                         0.000   
75%             0.000                         0.000   
max       4170901.000                        35.000   

       DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS  NO.OF_INQUIRIES  loan_default  
count                           233154.000       233154.000    233154.000  
mean                                 0.097            0.207         0.217  
std                                  0.384            0.706         0.412  
min                                  0.000            0.000         0.000  
25%                                  0.000            0.000         0.000  
50%                                  0.000            0.000         0.000  
75%                                  0.000            0.000         0.000  
max                                 20.000           36.000         1.000  

[8 rows x 35 columns]

In [5]:
raw_df.dtypes.sort_values(ascending=True)

UniqueID                                 int64
PRI.NO.OF.ACCTS                          int64
PRI.ACTIVE.ACCTS                         int64
PRI.OVERDUE.ACCTS                        int64
PRI.CURRENT.BALANCE                      int64
PRI.SANCTIONED.AMOUNT                    int64
PRI.DISBURSED.AMOUNT                     int64
SEC.NO.OF.ACCTS                          int64
NO.OF_INQUIRIES                          int64
SEC.ACTIVE.ACCTS                         int64
SEC.CURRENT.BALANCE                      int64
SEC.SANCTIONED.AMOUNT                    int64
SEC.DISBURSED.AMOUNT                     int64
PRIMARY.INSTAL.AMT                       int64
SEC.INSTAL.AMT                           int64
NEW.ACCTS.IN.LAST.SIX.MONTHS             int64
DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS      int64
SEC.OVERDUE.ACCTS                        int64
PERFORM_CNS.SCORE                        int64
loan_default                             int64
Driving_flag                             int64
disbursed_amo

In [6]:
loan = raw_df.loan_default.value_counts().to_frame().rename(columns={"loan_default":"absolute"})
loan["percent"] = (loan.apply(lambda x: x/x.sum()*100).round(2))
display(loan)

absolute  percent
0    182543    78.29
1     50611    21.71

Notice that the data is highly unbalanced and thus we will need to balance the data before it is fitted on to the model.

### Balance data

We shall balance the data by upsampling the minority group

In [7]:
print('Before Resampling:')
print(raw_df['loan_default'].value_counts())
df_majority = raw_df[raw_df['loan_default']==0]
df_minority = raw_df[raw_df['loan_default']==1]


df_minority_upsampled = resample(df_minority, 
                                 replace=True,                  # sample with replacement
                                 n_samples=len(df_majority),    # to match majority class
                                 random_state=123)              # reproducible results
df_upsampled = pd.concat([df_majority, df_minority_upsampled])
print('After Resampling:')
print(df_upsampled['loan_default'].value_counts())

Before Resampling:
0    182543
1     50611
Name: loan_default, dtype: int64
After Resampling:
0    182543
1    182543
Name: loan_default, dtype: int64


In [8]:
# Check for missing values
df_upsampled.isnull().sum().to_frame().rename(columns={0:"Missing Values"})

Missing Values
UniqueID                                          0
disbursed_amount                                  0
asset_cost                                        0
ltv                                               0
branch_id                                         0
supplier_id                                       0
manufacturer_id                                   0
Current_pincode_ID                                0
Date.of.Birth                                     0
Employment.Type                               11875
DisbursalDate                                     0
State_ID                                          0
Employee_code_ID                                  0
MobileNo_Avl_Flag                                 0
Aadhar_flag                                       0
PAN_flag                                          0
VoterID_flag                                      0
Driving_flag                                      0
Passport_flag                                     0
PERFORM_CNS.SCORE                                 0
PERFORM_CNS.SCORE.DESCRIPTION                     0
PRI.NO.OF.ACCTS                                   0
PRI.ACTIVE.ACCTS                                  0
PRI.OVERDUE.ACCTS                                 0
PRI.CURRENT.BALANCE                               0
PRI.SANCTIONED.AMOUNT                             0
PRI.DISBURSED.AMOUNT                              0
SEC.NO.OF.ACCTS                                   0
SEC.ACTIVE.ACCTS                                  0
SEC.OVERDUE.ACCTS                                 0
SEC.CURRENT.BALANCE                               0
SEC.SANCTIONED.AMOUNT                             0
SEC.DISBURSED.AMOUNT                              0
PRIMARY.INSTAL.AMT                                0
SEC.INSTAL.AMT                                    0
NEW.ACCTS.IN.LAST.SIX.MONTHS                      0
DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS               0
AVERAGE.ACCT.AGE                                  0
CREDIT.HISTORY.LENGTH                             0
NO.OF_INQUIRIES                                   0
loan_default                                      0

Employment type has missing values which need to be taken care of during pre-processing

### Data pre-processing 

In [9]:
def removeOutlier(dataSet, feature):    
    q1=dataSet[feature].quantile(0.25)
    q3=dataSet[feature].quantile(0.75)
    IQR=q3-q1
    lowerLimit = q1 - 1.5 * IQR
    UpperLimit = q3 + 1.5 * IQR 
    dataSet = dataSet[dataSet[feature]< UpperLimit]
    dataSet = dataSet[dataSet[feature]> lowerLimit]
    return dataSet

In [10]:
# Conver year, months to year float
def convert_years_months_to_float(yearmonth):
    split_per = yearmonth.split(' ')
    years = split_per[0][:-3]
    months = split_per[1][:-3]
    return int(years) + int(months)/12

In [11]:
# Get age from DOB and drop DOB. We are likely to get more insights from age as compared to DOB.
def calculateAge(birthDate):
    birthDate = datetime.strptime(birthDate, "%d-%m-%y").date()
    today = date.today()
    age = today.year - birthDate.year - ((today.month, today.day) < (birthDate.month, birthDate.day))
    return age

In [12]:
df_upsampled['Age'] = df_upsampled['Date.of.Birth'].apply(calculateAge)
df_upsampled = df_upsampled.drop('Date.of.Birth',axis=1)

df_upsampled['AVERAGE_ACCT_AGE'] = df_upsampled['AVERAGE.ACCT.AGE'].apply(convert_years_months_to_float)
df_upsampled = df_upsampled.drop('AVERAGE.ACCT.AGE',axis=1)

df_upsampled['CREDIT_HISTORY_LENGTH'] = df_upsampled['CREDIT.HISTORY.LENGTH'].apply(convert_years_months_to_float)
df_upsampled = df_upsampled.drop('CREDIT.HISTORY.LENGTH',axis=1)

We have observed that Employment type has missing values. We need to handle that.

In [13]:
df_upsampled['Employment.Type'] = df_upsampled['Employment.Type'].fillna("Missing")

In [14]:
df_upsampled.MobileNo_Avl_Flag.value_counts()

1    365086
Name: MobileNo_Avl_Flag, dtype: int64

In [32]:
df_upsampled['SEC.OVERDUE.ACCTS'].value_counts()

0    363048
1      1684
2       211
3        85
4        32
5        14
6        10
8         1
7         1
Name: SEC.OVERDUE.ACCTS, dtype: int64

In [15]:
#Correlation Matrix calculation
corr_mat = df_upsampled.corr()
corr_mat['loan_default'].sort_values(ascending = False)

loan_default                           1.000000
ltv                                    0.123027
disbursed_amount                       0.095516
State_ID                               0.058374
VoterID_flag                           0.052043
NO.OF_INQUIRIES                        0.047991
PRI.OVERDUE.ACCTS                      0.047494
DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS    0.040104
UniqueID                               0.038315
branch_id                              0.036023
Current_pincode_ID                     0.034228
supplier_id                            0.033599
Employee_code_ID                       0.027389
asset_cost                             0.016631
Age                                    0.008223
PAN_flag                               0.003469
SEC.OVERDUE.ACCTS                     -0.000975
SEC.INSTAL.AMT                        -0.003588
Driving_flag                          -0.005962
PRI.DISBURSED.AMOUNT                  -0.006191
PRI.SANCTIONED.AMOUNT                 -0

In [16]:
# We shall use the upsampled dataframe for model training
df = df_upsampled
df.head()

UniqueID  disbursed_amount  asset_cost    ltv  branch_id  supplier_id  \
0    420825             50578       58400  89.55         67        22807   
2    417566             53278       61360  89.63         67        22807   
5    518279             54513       61900  89.66         67        22807   
6    529269             46349       61500  76.42         67        22807   
7    510278             43894       61900  71.89         67        22807   

   manufacturer_id  Current_pincode_ID Employment.Type DisbursalDate  ...  \
0               45                1441        Salaried      03-08-18  ...   
2               45                1497   Self employed      01-08-18  ...   
5               45                1501   Self employed      19-09-18  ...   
6               45                1502        Salaried      23-09-18  ...   
7               45                1501        Salaried      16-09-18  ...   

   SEC.DISBURSED.AMOUNT  PRIMARY.INSTAL.AMT  SEC.INSTAL.AMT  \
0                     0                   0               0   
2                     0                   0               0   
5                     0                1347               0   
6                     0                   0               0   
7                     0                   0               0   

   NEW.ACCTS.IN.LAST.SIX.MONTHS  DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS  \
0                             0                                    0   
2                             0                                    0   
5                             0                                    0   
6                             0                                    0   
7                             0                                    0   

   NO.OF_INQUIRIES  loan_default  Age  AVERAGE_ACCT_AGE CREDIT_HISTORY_LENGTH  
0                0             0   37          0.000000              0.000000  
2                0             0   36          0.000000              0.000000  
5                0             0   31          1.750000              2.000000  
6                0             0   33          0.000000              0.000000  
7                0             0   32          0.166667              0.166667  

[5 rows x 41 columns]

### Drop columns that offer very little or no information in training the models

We need to drop columns that will not help the model
1. The UniqueID is only an identifier.
2. The MobileNo_Avl_Flag contains the same value all through.
3. DisbursalDate: Logically, we do not expect a lot from the date a loan was disbursed.
4. Employee_code_ID: The employee of the organization who logged the disbursement will not help much in predicting default.
5. SEC.OVERDUE.ACCTS: 99% of the values are 0. Very little information can be obtained.
6. PERFORM_CNS.SCORE.DESCRIPTION: This is a categorical version (description) of PERFORM_CNS.SCORE. We better work with the number.

In [17]:
columns_to_drop = ['UniqueID','MobileNo_Avl_Flag','DisbursalDate','SEC.OVERDUE.ACCTS', 'Employee_code_ID',\
                   'PERFORM_CNS.SCORE.DESCRIPTION', 'branch_id', 'supplier_id', 'manufacturer_id', 'Current_pincode_ID']
df=df.drop(columns=columns_to_drop)

In [18]:
df.dtypes.value_counts()

int64      27
float64     3
object      1
dtype: int64

#### Encode the categorical value Employment Type

In [19]:
df["Employment.Type"] = df["Employment.Type"].astype('category')
df["Employment.Type_cat"] = df["Employment.Type"].cat.codes
df = df.drop(columns=["Employment.Type"])

In [20]:
df.dtypes.value_counts()

int64      27
float64     3
int8        1
dtype: int64

#### Separate the predicted from predictors

In [21]:
X =df.drop('loan_default',axis=1)
y = df['loan_default']  

### Scaling Data
We shall use MinMax Scaler to scale the data and also take care of extreme values

In [22]:
min_max_scaler = preprocessing.MinMaxScaler()

In [23]:
X_minmax = min_max_scaler.fit_transform(X)
X_minmax

array([[0.03812527, 0.01344228, 0.93585972, ..., 0.        , 0.        ,
        0.5       ],
       [0.04088812, 0.01530158, 0.93680122, ..., 0.        , 0.        ,
        1.        ],
       [0.04215187, 0.01564078, 0.93715429, ..., 0.05691057, 0.05128205,
        1.        ],
       ...,
       [0.04471006, 0.01758112, 0.93232906, ..., 0.        , 0.        ,
        0.5       ],
       [0.03512809, 0.01761881, 0.75579616, ..., 0.02439024, 0.01923077,
        0.5       ],
       [0.03705083, 0.01354655, 0.93691891, ..., 0.02168022, 0.02136752,
        0.5       ]])

## Split data
We shall split the data to obtain a test size of 20%

In [24]:
X_train, X_test, y_train, y_test = train_test_split(X_minmax,y, test_size=0.2, random_state=0)

# Model Training and testing

### Logistic Regression Classifier

In [25]:
LRclf = LogisticRegression(C=1.0, class_weight=None,fit_intercept=True,max_iter=100)
LRclf.fit(X_train, y_train)
error = (1 - LRclf.score(X_test, y_test))*100
print('Score  = ',LRclf.score(X_test, y_test)*100, '%','\nError = %f' % error, '%')

Score  =  58.67183434221699 % 
Error = 41.328166 %


C:\Users\kgn71188\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [26]:
predictions = LRclf.predict(X_test)
TN, FP, FN, TP = confusion_matrix(y_test, predictions).ravel()

print('True Positive(TP)  = ', TP)
print('False Positive(FP) = ', FP)
print('True Negative(TN)  = ', TN)
print('False Negative(FN) = ', FN)

accuracy =  (TP+TN) /(TP+FP+TN+FN)

print('Accuracy of the binary classification = {:0.3f}'.format(accuracy))

True Positive(TP)  =  22930
False Positive(FP) =  16863
True Negative(TN)  =  19911
False Negative(FN) =  13314
Accuracy of the binary classification = 0.587


### Random Forest Classifier

In [27]:
RFclf = RandomForestClassifier()
RFclf.fit(X_train, y_train)
error = (1 - RFclf.score(X_test, y_test))*100
print('Score  = ',RFclf.score(X_test, y_test)*100, '%','\nError = %f' % error, '%')

Score  =  92.9674874688433 % 
Error = 7.032513 %


In [28]:
predictions = RFclf.predict(X_test)
TN, FP, FN, TP = confusion_matrix(y_test, predictions).ravel()

print('True Positive(TP)  = ', TP)
print('False Positive(FP) = ', FP)
print('True Negative(TN)  = ', TN)
print('False Negative(FN) = ', FN)

accuracy =  (TP+TN) /(TP+FP+TN+FN)

print('Accuracy of the binary classification = {:0.2f}%'.format(accuracy*100))

True Positive(TP)  =  34557
False Positive(FP) =  3448
True Negative(TN)  =  33326
False Negative(FN) =  1687
Accuracy of the binary classification = 92.97%


### Decision Tree Classifier

In [29]:
DTclf  = DecisionTreeClassifier()
DTclf.fit(X_train, y_train)
error = (1 - DTclf.score(X_test, y_test))*100
print('Score  = ',DTclf.score(X_test, y_test)*100, '%','\nError = %f' % error, '%')

Score  =  85.9007368046235 % 
Error = 14.099263 %


In [31]:
predictions = DTclf.predict(X_test)
TN, FP, FN, TP = confusion_matrix(y_test, predictions).ravel()

print('True Positive(TP)  = ', TP)
print('False Positive(FP) = ', FP)
print('True Negative(TN)  = ', TN)
print('False Negative(FN) = ', FN)

accuracy =  (TP+TN) /(TP+FP+TN+FN)

print('Accuracy of the binary classification = {:0.2f}%'.format(accuracy*100))

True Positive(TP)  =  34812
False Positive(FP) =  8863
True Negative(TN)  =  27911
False Negative(FN) =  1432
Accuracy of the binary classification = 85.90%


The Random Forest Classifier has done excellent. An accuracy of 93% is good enough. It also min